mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs

```diff
@@ -0,0 +1,1494 @@
+        hash: BlockHash,
+        time: u32,
+        sapling_tree: &[u8],
+    ) -> MmResult<(), ZcoinStorageError> {
+        let ticker = self.ticker.clone();
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let walletdb_account_table = db_transaction.table::<WalletDbBlocksTable>().await?;
+
+        // check if account exists
+        let maybe_min_account = walletdb_account_table
+            .cursor_builder()
+            .only("ticker", &ticker)?
+            .bound("height", 0u32, u32::MAX)
+            .open_cursor(WalletDbBlocksTable::TICKER_HEIGHT_INDEX)
+            .await?
+            .next()
+            .await?;
+        if maybe_min_account.is_some() {
+            return MmError::err(ZcoinStorageError::TableNotEmpty(
+                "Account table is not empty".to_string(),
+            ));
+        }
+
+        let block = WalletDbBlocksTable {
+            height: u32::from(height),
+            hash: hash.0.to_vec(),
+            time,
+            sapling_tree: sapling_tree.to_vec(),
+            ticker: ticker.clone(),
+        };
+        let walletdb_blocks_table = db_transaction.table::<WalletDbBlocksTable>().await?;
+        let height = u32::from(height);
+        let index_keys = MultiIndex::new(WalletDbBlocksTable::TICKER_HEIGHT_INDEX)
+            .with_value(&ticker)?
+            .with_value(num_to_bigint!(height)?)?;
+        walletdb_blocks_table
+            .replace_item_by_unique_multi_index(index_keys, &block)
+            .await?;
+
+        Ok(())
+    }
+}
+
+impl WalletIndexedDb {
+    pub async fn insert_block(
+        &self,
+        block_height: BlockHeight,
+        block_hash: BlockHash,
+        block_time: u32,
+        commitment_tree: &CommitmentTree<Node>,
+    ) -> MmResult<(), ZcoinStorageError> {
+        let ticker = self.ticker.clone();
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let walletdb_blocks_table = db_transaction.table::<WalletDbBlocksTable>().await?;
+
+        let mut encoded_tree = Vec::new();
+        commitment_tree.write(&mut encoded_tree).unwrap();
+
+        let hash = &block_hash.0[..];
+        let block = WalletDbBlocksTable {
+            height: u32::from(block_height),
+            hash: hash.to_vec(),
+            time: block_time,
+            sapling_tree: encoded_tree,
+            ticker: ticker.clone(),
+        };
+
+        let index_keys = MultiIndex::new(WalletDbBlocksTable::TICKER_HEIGHT_INDEX)
+            .with_value(&ticker)?
+            .with_value(u32::from(block_height))?;
+
+        Ok(walletdb_blocks_table
+            .replace_item_by_unique_multi_index(index_keys, &block)
+            .await
+            .map(|_| ())?)
+    }
+
+    pub async fn get_balance(&self, account: AccountId) -> MmResult<Amount, ZcoinStorageError> {
+        let ticker = self.ticker.clone();
```

Comment:
We don't need to clone this, as it's only passed by reference. In this module we have too many extra cloning like this, I don't want to point them all one by one, so I leave the rest to you to fix them.

Reply:
all done